core/codenvy-hosted-api-workspace/src/main/java/com/codenvy/api/workspace/server/stack/StackPermissionStorage.java

```diff
@@ -0,0 +1,135 @@
+
+/**
+ * Implementation of {@link PermissionsStorage} for storing permissions of {@link StackDomain}
+ *
+ * <p>This implementation adapts {@link Permissions} and {@link AclEntry} and use
+ * {@link StackDao} as storage of permissions
+ *
+ * @author Sergii Leschenko
+ */
+@Singleton
+public class StackPermissionStorage implements PermissionsStorage {
+    private final StackDao stackDao;
+
+    @Inject
+    public StackPermissionStorage(StackDao stackDao) {
+        this.stackDao = stackDao;
+    }
+
+    @Override
+    public Set<PermissionsDomain> getDomains() {
+        return ImmutableSet.of(new StackDomain());
+    }
+
+    @Override
+    public void store(PermissionsImpl permissions) throws ServerException, NotFoundException {
+        final StackImpl stack = stackDao.getById(permissions.getInstance());
+        stack.getAcl().removeIf(aclEntry -> aclEntry.getUser().equals(permissions.getUser()));
+        stack.getAcl().add(new AclEntryImpl(permissions.getUser(),
+                                            permissions.getActions()));
+        stackDao.update(stack);
+    }
+
+    @Override
+    public PermissionsImpl get(String user, String domain, String instance) throws ServerException, NotFoundException {
+        final StackImpl recipe = stackDao.getById(instance);
+        final Optional<AclEntryImpl> result = recipe.getAcl()
+                                                    .stream()
+                                                    .filter(aclEntry -> aclEntry.getUser().equals(user))
+                                                    .findAny();
+
+        if (!result.isPresent()) {
+            throw new NotFoundException(String.format("Permissions for user '%s' and instance '%s' of domain '%s' was not found",
+                                                      user, instance, domain));
+        }
+
+        return toPermission(instance, result.get());
+    }
+
+    @Override
+    public List<PermissionsImpl> getByInstance(String domain, String instance) throws ServerException, NotFoundException {
+        final StackImpl recipe = stackDao.getById(instance);
+        return toPermissions(instance, recipe.getAcl());
+    }
+
+    @Override
+    public boolean exists(String user, String domain, String instance, String action) throws ServerException {
+        StackImpl recipe;
+        try {
+            recipe = stackDao.getById(instance);
+        } catch (NotFoundException e) {
+            return false;
+        }
+
+        final Set<String> allowedActions = recipe.getAcl()
+                                                 .stream()
+                                                 .filter(entry -> entry.getUser().equals(user) || entry.getUser().equals("*"))
+                                                 .flatMap(entry -> entry.getActions().stream())
+                                                 .collect(Collectors.toSet());
+        return allowedActions.contains(action);
```

Comment:
I'd say that `anyMatch` is better, as you don't need to allocate memory for `Set` + it is cheaper in the case of time.
```java
return recipe.getAcl()
             .stream()
             .filter(entry -> entry.getUser().equals(user) || entry.getUser().equals("*")) 
             .flatMap(entry -> entry.getActions().stream())
             .anyMatch(action::equals);
```